src/SlackFunction.ts

```diff
@@ -46,83 +113,412 @@ export class SlackFunction {
+    // cast to string array for convenience
+    const validConstraintsKeysAsStrings = validConstraintsKeys as string[];
+
+    errorIfInvalidConstraintKeys(constraints, validConstraintsKeysAsStrings, handler);
+
+    this.interactivityHandlers.push({ constraints, handler });
+    return this;
   }
 
-  /* Utility */
+  /**
+   * Attach a view_submission or view_closed interactivity handler
+   * to your SlackFunction
+   *
+   * ```
+   * Example:
+   * const viewHandler = async () => {};
+   * const viewHandler1 = async () => {};
+   * myFunc.view("id", viewHandler).view("id1", viewHandler1)
+   * ```
+   *
+   * @param callbackIdOrConstraints Provide a `callback_id` string
+   * a constraint object of type ViewConstraints
+   *
+   * ```
+   * Example:
+   * myFunc.view({ type: "view_submission" });
+   * myFunc.view({ callback_id: "id", }, viewHandler)
+   * ```
+   *
+   * @param handler Provide a handler function
+   * @returns SlackFunction instance
+   */
+  public view(
+    callbackIdOrConstraints: string | RegExp | ViewConstraints,
+    handler: Middleware<SlackViewMiddlewareArgs<SlackViewAction>>,
+  ): this {
+    // normalize constraints
+    const constraints: ViewConstraints = (
+      typeof callbackIdOrConstraints === 'string' ||
+      util.types.isRegExp(callbackIdOrConstraints)
+    ) ?
+      { callback_id: callbackIdOrConstraints, type: 'view_submission' } :
+      callbackIdOrConstraints;
+
+    // declare our valid constraints keys
+    const validConstraintsKeys: ViewConstraintsKeys = ['callback_id', 'type'];
+    // cast to string array for convenience
+    const validConstraintsKeysAsStrings = validConstraintsKeys as string[];
+
+    errorIfInvalidConstraintKeys(constraints, validConstraintsKeysAsStrings, handler);
+
+    this.interactivityHandlers.push({ constraints, handler });
+    return this;
+  }
+
+  private matchesFuncConstraints(args: AnyMiddlewareArgs): boolean {
+    if ('function' in args.payload) {
+      return this.callbackId === args.payload.function.callback_id;
+    }
+    return false;
+  }
+
+  /**
+   * Returns a a single middleware to global
+   * middleware chain. Responsible for returning
+   * handlers required for either function execution
+   * or function interactivity event handlering to
+   * the global event processing chain.
+  */
   public getMiddleware(): Middleware<AnyMiddlewareArgs> {
     return async (args): Promise<void> => {
-      if (isFunctionExecutedEvent(args) && this.matchesConstraints(args)) {
-        return this.run(args);
+      // handle function executed event
+      if ((isFunctionExecutedEvent(args) && this.matchesFuncConstraints(args))) {
+        try {
+          this.logger.debug('🚀 Executing my main handler:', this.handler);
+          return await this.runHandler(args);
+        } catch (error) {
+          this.logger.error('⚠️ Something went wrong executing:', this.handler);
```

Comment:
I wanted to check what this log looks like but I couldn't trip it.

I edited the function body to throw an error, then this is the output I see:

```
[DEBUG]  socket-mode:SocketModeClient:0 Received a message on the WebSocket: {"envelope_id": ... very long payload here truncated}
[DEBUG]  socket-mode:SocketModeClient:0 Calling ack() - type: events_api, envelope_id: 523882e6-c036-43ba-8e7a-e5c15c7f7007, data: undefined
[DEBUG]  socket-mode:SocketModeClient:0 send() method was called in state: connected, state hierarchy: connected,ready
[DEBUG]  socket-mode:SocketModeClient:0 Sending a WebSocket message: {"envelope_id":"523882e6-c036-43ba-8e7a-e5c15c7f7007","payload":{}}
[DEBUG]  bolt-app initialized
[DEBUG]  bolt-app No conversation ID for incoming event
[DEBUG]  SlackFunction: [review_approval] 🚀 Executing my main handler: [AsyncFunction: notifyApprover]
/Users/fmaj/src/slapp-BoltJSHermes/listeners/functions/request-approval.js:12
  throw new Error('boomsies');
        ^

Error: boomsies
    at SlackFunction.notifyApprover [as handler] (/Users/fmaj/src/slapp-BoltJSHermes/listeners/functions/request-approval.js:12:9)
    at SlackFunction.runHandler (/Users/fmaj/src/bolt-js/dist/SlackFunction.js:60:18)
    at Array.<anonymous> (/Users/fmaj/src/bolt-js/dist/SlackFunction.js:184:39)
    at invokeMiddleware (/Users/fmaj/src/bolt-js/dist/middleware/process.js:12:53)
    at next (/Users/fmaj/src/bolt-js/dist/middleware/process.js:13:29)
    at Array.<anonymous> (/Users/fmaj/src/bolt-js/dist/conversation-store.js:67:15)
    at invokeMiddleware (/Users/fmaj/src/bolt-js/dist/middleware/process.js:12:53)
    at Object.next (/Users/fmaj/src/bolt-js/dist/middleware/process.js:13:29)
    at Array.<anonymous> (/Users/fmaj/src/bolt-js/dist/middleware/builtin.js:273:20)
    at invokeMiddleware (/Users/fmaj/src/bolt-js/dist/middleware/process.js:12:53)
bolt-app local run exited with code 1
```

Note also that this kills the entire process. Pretty weird as I would assume this try/catch block would catch it. Guess not. Is this expected? I would have expected to see "Something went wrong", and possibly also for the process not to die. Whatever the expectation is, let's make sure to add a test for this behaviour if we can.

Reply:
Yeah, so the `return await` ing in the line you highlighted looks right, but I missed an `await` in runHandler, so that was immediately fulfilling with undefined, and then getting returned out all the way up, bypassing my well-intentioned catch block. I've simplified code little bit and I think it's much clearer now. Added a test for the case as well.

![Screen Shot 2022-08-29 at 4 20 48 PM](https://user-images.githubusercontent.com/55667998/187315387-f5540236-585a-4da1-a90d-56476cf6358a.png)